Expose the current game's state via WebSockets

LeNitrous opened the pull request:
- Addresses https://github.com/ppy/osu/discussions/13011

Adds a new `GameStateBroadcaster` to the desktop client which listens to incoming `WebSocket` requests and sends a JSON payload to the other receiving ends whenever the game's state is changed via bindables which are throttled to avoid spamming.

This will allow external API consumers such as integration to their programmable peripherals and streamer overlays to connect to the `WebSocket` and see the current state of the game.

TODO:
- [x] Expose settings via UI or by adding a new config file.

Demonstration:

https://user-images.githubusercontent.com/20495991/113465916-2bf73e80-946a-11eb-91dc-57510272ac87.mp4

LeNitrous committed:
implement GameStateBroadcaster

frenzibyte commented:
Is there a reason you PR'd this even though it's still WIP? would usually expect the author wanting to discuss a certain piece of the implementation on that case.

peppy commented on osu.Desktop/GameStateBroadcaster.cs:
```diff
@@ -0,0 +1,187 @@
+
+        public GameStateBroadcaster()
+        {
+            listener.Prefixes.Add("http://localhost:8080/");
```
We obviously can't use such a common port.

LeNitrous replied:
Would it be fine if this can be replaced in settings?

peppy replied:
No, it should be well defined and hopefully follow existing standards for local port assignments (ensuring we don't collide with other apps). Will need to do some reading and provide references to how this was decided.

LeNitrous replied:
In the meantime I've replaced the port number to `7270`.

LeNitrous commented:
In the TODO, I'm unsure whether we'd want this to be under the Settings UX or make the user create a config file. Additionally, there are a few more things needed to expose such as score, stats, and mods of the currently played beatmap.

peppy commented:
I think you mean UI, not UX. We do not rely on configuration files at all, so yes it needs to have a toggle in the settings. Just a single one to enable or disable.

peppy commented on osu.Desktop/GameStateBroadcaster.cs:
```diff
@@ -0,0 +1,187 @@
+                var client = new WebSocketClient(wsContext.WebSocket);
+                clients.Add(client);
+
+                var thread = new Thread(() => client.Handle().Wait())
```
All this `new Thread` stuff cannot exist. Please read up on newer .NET paradigms and follow them, as I'm 12020% sure that this is not the correct way.

bdach commented:
I very strongly feel this should be off by default.

peppy commented:
Yep, defaulting to off is a good idea for now.

LeNitrous committed:
refactor broadcaster component

+ change port to something less common
+ add a toggle in settings
+ add a config entry

LeNitrous committed:
ensure there is no broadcast before closing all connections

LeNitrous committed:
refactor more

+ move GameState to its own class
+ make GameState and GameStateBroadcaster internal

LeNitrous committed:
a bit more refactoring

+ move bindables to GameState
+ make custom converters

LeNitrous commented:
Not sure how to access `ScoreProcessor` or any information inside `Player` as that's needed left but for now I'll move this as ready for review since I'm fairly comfortable at its current state.

bdach commented:
CI disagrees (several unresolved inspections).

LeNitrous committed:
fix appveyor issues

LeNitrous committed:
additional fixes

+ ensure newly connected clients receive a payload
+ fix user activity being null

LeNitrous committed:
ensure lock to fix appveyor

LeNitrous commented:
I'm probably sure code quality on this has been thrown out the window at some point. Not sure how to make it any better either.

peppy commented on osu.Desktop/GameState.cs:
```diff
@@ -0,0 +1,103 @@
+            }
+        }
+
+        private class WorkingBeatmapConverter : JsonConverter<Bindable<WorkingBeatmap>>
+        {
+            public override Bindable<WorkingBeatmap> ReadJson(JsonReader reader, Type objectType, Bindable<WorkingBeatmap> existingValue, bool hasExistingValue, JsonSerializer serializer)
+            {
+                throw new NotImplementedException();
+            }
+
+            public override void WriteJson(JsonWriter writer, Bindable<WorkingBeatmap> value, JsonSerializer serializer)
+            {
+                var beatmap = value.Value;
+                serializer.Serialize(writer, beatmap.BeatmapInfo);
+            }
+        }
+
+        private class UserActivityConverter : JsonConverter<Bindable<UserActivity>>
+        {
+            public override Bindable<UserActivity> ReadJson(JsonReader reader, Type objectType, Bindable<UserActivity> existingValue, bool hasExistingValue, JsonSerializer serializer)
+            {
+                throw new NotImplementedException();
+            }
+
+            public override void WriteJson(JsonWriter writer, Bindable<UserActivity> value, JsonSerializer serializer)
+            {
+                string status = null;
+
+                switch (value.Value)
+                {
+                    case UserActivity.Modding _:
+                        status = "modding";
```
this is not how you do json. please look at any other example of json output anywhere in the project.

peppy commented on osu.Desktop/GameStateBroadcaster.cs:
```diff
@@ -0,0 +1,263 @@
+
+            enabled.BindValueChanged(state =>
+            {
+                if (state.NewValue)
+                    Task.Factory.StartNew(start, TaskCreationOptions.LongRunning);
+                else
+                    stop();
```
there's no protection against start being called while stop is still running.

stops task is hidden inside a private

peppy commented on osu.Desktop/GameStateBroadcaster.cs:
```diff
@@ -0,0 +1,263 @@
+            cancellationToken = new CancellationTokenSource();
+            listener.Start();
+
+            while (!cancellationToken.IsCancellationRequested)
```
is this taken from an example elsewhere?

LeNitrous replied:
It's taken from [this article](https://mcguirev10.com/2019/01/15/simple-multi-client-websocket-server.html). Many MSDN articles point towards using a Kestrel Server which is only available on ASP.NET projects. Though I've only used it as a rough basis to understand the lifecycle of a websocket.

peppy commented on osu.Desktop/GameStateBroadcaster.cs:
```diff
@@ -0,0 +1,263 @@
+            public void Start()
+            {
+                OnStart?.Invoke(this);
+                Task.WhenAll(receive(), send()).ConfigureAwait(false);
```
Likewise, are you following some example for this implementation? If so, a reference would be appreciated (MSDN etc.)

LeNitrous committed:
add package dependencies

LeNitrous committed:
Rewrite GameStateBroadcaster to use Kestrel instead of HttpListener

LeNitrous committed:
move to main project

LeNitrous committed:
add tests

LeNitrous committed:
some fixes

+ expose clients property as a read-only list
+ move web host building to start method
+ ensure config bindable is disabled while stopping
+ stop server on disposal

LeNitrous committed:
move back to desktop project

this is due to the Kestrel package being outdated (see: https://github.com/dotnet/aspnetcore/issues/28112)

LeNitrous committed:
ensure a payload is sent after client connects

LeNitrous committed:
remove unused public properties

+ remove IsListening in favor of nuill checking host instead
+ remove Clients as it is unused

LeNitrous committed:
hide methods related to adding and removing clients

There is no real reason to expose them to the public. There is also no real need to pass the entire drawable to the web host.

LeNitrous committed:
refactor GameState

LeNitrous committed:
pass InspectCode

LeNitrous commented:
As discussed, I have rewritten the websocket server to use `Microsoft.AspNetCore.Server.Kestrel`. Due to the `Microsoft.AspNetCore.App` framework being incompatible with .NET Standard projects, this is only available to the desktop project which uses .NET 5.0.

This PR only exposes the current ruleset, current beatmap and current selected mods. Other things that need to be exposed are the current play's statistics but that will require a separate PR as there are things under `Player` that need to be exposed and is probably out of scope for this PR.

peppy commented:
If possible can you attach the test code you had? I'd like to have this tested so we will need to figure out a way to make that happen still.

LeNitrous commented:
I think adding a reference to `osu.Desktop` inside `osu.Game.Tests` would probably be fine right? I'll add that back in if that's fine.

LeNitrous committed:
remove comment

LeNitrous committed:
make broadcaster available in tests

LeNitrous commented:
Added tests as requested.

peppy commented on osu.Game.Tests/osu.Game.Tests.csproj:
```diff
@@ -21,5 +21,6 @@
     <ProjectReference Include="..\osu.Game.Rulesets.Catch\osu.Game.Rulesets.Catch.csproj" />
     <ProjectReference Include="..\osu.Game.Rulesets.Mania\osu.Game.Rulesets.Mania.csproj" />
     <ProjectReference Include="..\osu.Game.Rulesets.Taiko\osu.Game.Rulesets.Taiko.csproj" />
+    <ProjectReference Include="..\osu.Desktop\osu.Desktop.csproj" />
```
This can't really exist, which is why I was asking for the test to be provided in non-committed format. But it's fine to leave like this for now, we'll need to figure out the correct path forward here.

peppy committed:
Merge branch 'master' into publish-game-state

peppy commented:
Test immediately fails for me:

![20211018 175910 (dotnet)](https://user-images.githubusercontent.com/191335/137700763-44b246d2-2508-4720-9885-baacc65edf46.png)

```csharp

 8573: TestSceneGameStateBroadcaster
[runtime] 2021-10-18 08:58:21 [error]: An unobserved error has occurred.
[runtime] 2021-10-18 08:58:21 [error]: System.AggregateException: A Task's exception(s) were not observed either by Waiting on the Task or accessing its Exception property. As a result, the unobserved exception was rethrown by the finalizer thread. (Object reference not set to an instance of an object.)
[runtime] 2021-10-18 08:58:21 [error]: ---> System.NullReferenceException: Object reference not set to an instance of an object.
[runtime] 2021-10-18 08:58:21 [error]: at osu.Desktop.GameStateBroadcaster.<stop>b__20_0() in /Users/dean/Projects/osu/osu.Desktop/GameStateBroadcaster.cs:line 165
[runtime] 2021-10-18 08:58:21 [error]: --- End of inner exception stack trace ---
[runtime] 2021-10-18 08:58:21 [error]: An unobserved error has occurred. (inner)
[runtime] 2021-10-18 08:58:21 [error]: System.NullReferenceException: Object reference not set to an instance of an object.
[runtime] 2021-10-18 08:58:21 [error]: at osu.Desktop.GameStateBroadcaster.<stop>b__20_0() in /Users/dean/Projects/osu/osu.Desktop/GameStateBroadcaster.cs:line 165
 8774: step 1 
 8975: step 2 BroadcasterAlive

```

peppy commented:
Don't get ahead of yourself. The goal here is to great the basic communication structure.

LeNitrous committed:
fix possible null reference

LeNitrous committed:
use `IBindable<T>`s instead

overlisted commented:
Hey, is this PR dead? I would absolutely love to see this feature implemented

peppy commented:
It is not closed, therefore not "dead". Please refrain from commenting unless you are reviewing or adding to it, though.
